processor/deltatocumulativeprocessor/internal/data/add.go

```diff
@@ -12,120 +12,133 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/putil/pslice"
 )
 
+// Aggregator performs an operation on two datapoints.
+// Given [pmetric] types are mutable by nature, this logically works as follows:
+//
+//	*state = op(state, dp)
+//
+// See [Adder] for an implementation.
+type Aggregator interface {
+	Numbers(state, dp pmetric.NumberDataPoint) error
+	Histograms(state, dp pmetric.HistogramDataPoint) error
+	Exponential(state, dp pmetric.ExponentialHistogramDataPoint) error
+}
+
+var _ Aggregator = (*Adder)(nil)
+
+// Adder adds (+) datapoints.
+type Adder struct{}
+
 var maxBuckets = 160
 
-func (dp Number) Add(in Number) Number {
-	switch in.ValueType() {
+func (add Adder) Numbers(state, dp pmetric.NumberDataPoint) error {
+	switch dp.ValueType() {
 	case pmetric.NumberDataPointValueTypeDouble:
-		v := dp.DoubleValue() + in.DoubleValue()
-		dp.SetDoubleValue(v)
+		v := state.DoubleValue() + dp.DoubleValue()
+		state.SetDoubleValue(v)
 	case pmetric.NumberDataPointValueTypeInt:
-		v := dp.IntValue() + in.IntValue()
-		dp.SetIntValue(v)
+		v := state.IntValue() + dp.IntValue()
+		state.SetIntValue(v)
 	}
-	dp.SetTimestamp(in.Timestamp())
-	return dp
+	state.SetTimestamp(dp.Timestamp())
```

Comment:
What would define that "adding" two data points sets the timestamp to the newer one? I think the abstraction fails here a little bit, but I don't have a clear solution.

Reply:
oh good catch .. we could move that to delta.Aggregator which arguably fits this a lot better